spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/AbstractConnectionFactory.java

```diff
@@ -595,8 +595,10 @@ else if (soTimeout > 0) {
 								logger.warn("Timing out TcpNioConnection " +
 										    connection.getConnectionId());
 							}
-							connection.publishConnectionExceptionEvent(new SocketTimeoutException("Timing out connection"));
+							SocketTimeoutException exception = new SocketTimeoutException("Timing out connection");
+							connection.publishConnectionExceptionEvent(exception);
 							connection.timeout();
+							connection.sendExceptionToListener(exception);
```

Comment:
Sorry, I don't understand the question; we're just communicating the timeout to the waiting thread so the exception can be propagated to the caller. It doesn't matter if it happens before or after the socket is closed after the timeout detection.

Let me know if I'm missing something.